db/commitlog/commitlog.cc

```diff
@@ -1529,6 +1529,10 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
     if (!cfg.allow_going_over_size_limit && max_disk_size != 0 && totals.total_size_on_disk >= max_disk_size) {
         clogger.debug("Disk usage ({} MB) exceeds maximum ({} MB) - allocation will wait...", totals.total_size_on_disk/(1024*1024), max_disk_size/(1024*1024));
         auto f = cfg.reuse_segments ? _recycled_segments.not_empty() :  _disk_deletions.get_shared_future();
+        if (!f.available()) {
+            _new_counter = 0; // zero this so timer task does not duplicate the below flush
+            flush_segments(0); // force memtable flush already
+        }
         return f.handle_exception([this](auto ep) {
```

Comment:
I think we should avoid a flush if we've requested one already and it isn't complete (or is just partially complete, having flushed a few memtables), so that we don't create lots of small sstables. Do we have a notification about when a flush ends, or do we just get replay position releases?

Reply:
This is a bit risky, because we can't be sure that a flush call _will_ actually result in stuff really being released. In the case of memtables, we should probably be safe, since at most we should eventually get _more_ stuff released than we asked for. 
We could maybe replace the _new_counter check/set here and on_timer with simply having a slightly fuzzy "flush active" bool, which resets at first callback that actually frees data (since we cannot know how many there will be). Maybe per cf_id...

Reply:
No, it will not work. The problem is that memtables can flush themselves without us asking for it. So we could be notified of data being released before we add a bookkeep var, and then never receive the flush we asked for since memtable is already past the RP. So we'd end up thinking stuff is pending endlessly. 

I think keeping it like it is above is safest + good enough. We will only get here when we've exhausted segments from replenisher, which means nothing will enter this code until we've at least satisfies releasing one segment. Since this typically will release _more_ than   one segment, and we should not exhaust next segment until a "little later", we should not hit any congestion here.

Reply:
Ok. It could be improved, but I think we need to simplify commitlog before we complexify it even more.